services/continuousquery/service.go

```diff
@@ -0,0 +1,272 @@
+
+	// If no rp is specified, use the default one.
+	if cq.q.Target.Measurement.RetentionPolicy == "" {
+		cq.q.Target.Measurement.RetentionPolicy = dbi.DefaultRetentionPolicy
+	}
+
+	// Get the group by interval and report time for cq service.
+	interval, _ := cq.q.GroupByInterval()
+	// if interval < 5 minutes, set s.ReportInterval to 5 minutes.
+	if interval < DefaultReportTime {
+		s.ReportInterval = DefaultReportTime
+	} else {
+		s.ReportInterval = interval
+	}
+
+	// Get the group by offset.
+	offset, _ := cq.q.GroupByOffset()
+
+	// Check if the CQ should be run.
+	ok, nextRun := cq.shouldRunContinuousQuery(now, interval)
+	if !ok {
+		return false, err
+	}
+
+	// Calculate and set the time range for the query.
+	// startTime should be earlier than current time.
+	startTime := nextRun.Add(-interval - offset - 1).Truncate(interval).Add(offset)
+	endTime := startTime.Add(interval - offset).Truncate(interval).Add(offset)
+
+	if err := cq.q.SetTimeRange(startTime, endTime); err != nil {
+		return false, fmt.Errorf("unable to set time range: %s", err)
+	}
+
+	// execute the query and write the results
+	res := s.runContinuousQueryAndWriteResult(cq)
+	if res.Err != nil {
+		return false, res.Err
+	}
+
+	// update cq.LastRun and s.lastRuns
+	cq.LastRun = now.Truncate(interval)
+	s.lastRuns[cqi.Name] = cq.LastRun
+
+	return true, nil
+}
+
+// NewContinuousQuery returns a ContinuousQuery object with a parsed SQL statement.
+func NewContinuousQuery(database string, cqi *meta.ContinuousQueryInfo) (*ContinuousQuery, error) {
+	// Parse the query.
+	p := influxql.NewParser(strings.NewReader(cqi.Query))
+	defer p.Release()
+
+	YyParser := influxql.NewYyParser(p.GetScanner(), p.GetPara())
+	YyParser.ParseTokens()
+
+	qr, err := YyParser.GetQuery()
+	if err != nil {
+		return nil, err
+	}
+	if len(qr.Statements) == 0 {
+		return nil, nil
+	}
+	stmt := qr.Statements[0]
+
+	// Check if the statement is a valid continuous query.
+	q, ok := stmt.(*influxql.CreateContinuousQueryStatement)
+	if !ok || q.Source.Target == nil || q.Source.Target.Measurement == nil {
+		// q.Source.Target (destination) for the result of a SELECT INTO query.
+		return nil, errors.New("invalid continuous query")
+	}
+
+	// Create a new ContinuousQuery object.
+	cq := &ContinuousQuery{
+		Database: database,
+		Info:     cqi,
+		q:        q.Source,
+	}
+
+	return cq, nil
+}
+
+// shouldRunContinuousQuery returns true if the CQ should run.
+func (cq *ContinuousQuery) shouldRunContinuousQuery(now time.Time, interval time.Duration) (bool, time.Time) {
+	// Determine if we should run the continuous query based on the last time it ran.
+	if cq.HasRun { // The query has run before.
+		// Return the nextRun time for cq.
+		nextRun := cq.LastRun.Add(interval).Truncate(interval)
+		if nextRun.UnixNano() <= now.UnixNano() {
+			return true, nextRun
+		}
+		return false, cq.LastRun
+	} else { // The query never ran before.
+		// Retrieve the location from the CQ.
```

Comment:
![image](https://github.com/openGemini/openGemini/assets/108262220/3fd59a3d-a3df-4732-a3f3-85f7fe4700a9)
Else can be removed to simplify the code.